File: ReadyToShare/sMath/Java/sMath.java
package jtrsoftware.math;

import java.io.*;

public class SMath {
    public static char decimalSeparator = '.';

    public static boolean isNegative(String sV) {
        return sV != null && sV.startsWith("-");
    }

    public static String getAbs(String sV) {
        return isNegative(sV) ? sV.substring(1) : sV;
    }

    public static String cleanNumber(String sV) {
        boolean neg = isNegative(sV);
        String val = getAbs(sV);
        if (val == null || val.isEmpty()) return "0";

        StringBuilder sb = new StringBuilder(val);
        while (sb.length() > 1 && sb.charAt(0) == '0' && sb.charAt(1) != decimalSeparator) {
            sb.deleteCharAt(0);
        }

        int dotPos = sb.indexOf(String.valueOf(decimalSeparator));
        if (dotPos >= 0) {
            while (sb.length() > dotPos + 1 && sb.charAt(sb.length() - 1) == '0') {
                sb.deleteCharAt(sb.length() - 1);
            }
            if (sb.charAt(sb.length() - 1) == decimalSeparator) {
                sb.deleteCharAt(sb.length() - 1);
            }
        }

        String res = sb.toString();
        if (res.isEmpty() || res.equals(String.valueOf(decimalSeparator))) res = "0";

        return (neg && !res.equals("0")) ? "-" + res : res;
    }

    private static String[] alignStrings(String sA, String sB) {
        int dotA = sA.indexOf(decimalSeparator);
        int dotB = sB.indexOf(decimalSeparator);
        int decA = (dotA < 0) ? 0 : sA.length() - dotA - 1;
        int decB = (dotB < 0) ? 0 : sB.length() - dotB - 1;
        StringBuilder sbA = new StringBuilder(sA);
        StringBuilder sbB = new StringBuilder(sB);
        while (decA < decB) {
            if (dotA < 0) { sbA.append(decimalSeparator); dotA = sbA.length() - 1; }
            sbA.append('0'); decA++;
        }
        while (decB < decA) {
            if (dotB < 0) { sbB.append(decimalSeparator); dotB = sbB.length() - 1; }
            sbB.append('0'); decB++;
        }
        int intA = (dotA < 0) ? sbA.length() : dotA;
        int intB = (dotB < 0) ? sbB.length() : dotB;
        while (intA < intB) { sbA.insert(0, '0'); intA++; }
        while (intB < intA) { sbB.insert(0, '0'); intB++; }
        return new String[]{sbA.toString(), sbB.toString()};
    }

    public static String sum(String sA, String sB) {
        if (isNegative(sA) && isNegative(sB)) return "-" + internalSum(getAbs(sA), getAbs(sB));
        if (isNegative(sA)) return internalSub(getAbs(sB), getAbs(sA));
        if (isNegative(sB)) return internalSub(getAbs(sA), getAbs(sB));
        return cleanNumber(internalSum(sA, sB));
    }

    private static String internalSum(String sA, String sB) {
        String[] aligned = alignStrings(sA, sB);
        sA = aligned[0]; sB = aligned[1];
        StringBuilder res = new StringBuilder();
        int carry = 0;
        for (int i = sA.length() - 1; i >= 0; i--) {
            if (sA.charAt(i) == decimalSeparator) { res.insert(0, decimalSeparator); continue; }
            int dSum = (sA.charAt(i) - '0') + (sB.charAt(i) - '0') + carry;
            carry = dSum / 10;
            res.insert(0, (char)((dSum % 10) + '0'));
        }
        if (carry > 0) res.insert(0, carry);
        return res.toString();
    }

    public static String sub(String sA, String sB) {
        if (isNegative(sB)) return sum(sA, getAbs(sB));
        if (isNegative(sA)) return "-" + internalSum(getAbs(sA), sB);
        return cleanNumber(internalSub(sA, sB));
    }

    private static String internalSub(String sA, String sB) {
        String[] aligned = alignStrings(sA, sB);
        sA = aligned[0]; sB = aligned[1];
        boolean neg = false;
        if (sA.compareTo(sB) < 0) { String tmp = sA; sA = sB; sB = tmp; neg = true; }
        StringBuilder res = new StringBuilder();
        int borrow = 0;
        for (int i = sA.length() - 1; i >= 0; i--) {
            if (sA.charAt(i) == decimalSeparator) { res.insert(0, decimalSeparator); continue; }
            int diff = (sA.charAt(i) - '0') - (sB.charAt(i) - '0') - borrow;
            if (diff < 0) { diff += 10; borrow = 1; } else borrow = 0;
            res.insert(0, (char)(diff + '0'));
        }
        String sRes = cleanNumber(res.toString());
        return neg ? "-" + sRes : sRes;
    }

    public static String multiply(String sA, String sB) {
        boolean resNeg = isNegative(sA) ^ isNegative(sB);
        sA = getAbs(sA); sB = getAbs(sB);
        int dotA = sA.indexOf(decimalSeparator);
        int decA = (dotA < 0) ? 0 : sA.length() - dotA - 1;
        String intA = sA.replace(String.valueOf(decimalSeparator), "");
        int dotB = sB.indexOf(decimalSeparator);
        int decB = (dotB < 0) ? 0 : sB.length() - dotB - 1;
        String intB = sB.replace(String.valueOf(decimalSeparator), "");
        int totalDecs = decA + decB;
        String finalRes = "0";
        for (int i = intB.length() - 1; i >= 0; i--) {
            int digitB = intB.charAt(i) - '0';
            if (digitB == 0) continue;
            StringBuilder currentRes = new StringBuilder();
            int carry = 0;
            for (int j = intA.length() - 1; j >= 0; j--) {
                int digitA = intA.charAt(j) - '0';
                int prod = (digitA * digitB) + carry;
                carry = prod / 10;
                currentRes.insert(0, (char)((prod % 10) + '0'));
            }
            if (carry > 0) currentRes.insert(0, carry);
            for(int k=0; k < (intB.length() - 1 - i); k++) currentRes.append('0');
            finalRes = internalSum(finalRes, currentRes.toString());
        }
        if (totalDecs > 0) {
            while (finalRes.length() <= totalDecs) finalRes = "0" + finalRes;
            finalRes = new StringBuilder(finalRes).insert(finalRes.length() - totalDecs, decimalSeparator).toString();
        }
        String result = resNeg ? "-" + finalRes : finalRes;
        return cleanNumber(result);
    }

    public static String divide(String sA, String sB) {
        if (cleanNumber(sB).equals("0")) return "0";
        boolean resNeg = isNegative(sA) ^ isNegative(sB);
        sA = getAbs(sA); sB = getAbs(sB);
        String[] aligned = alignStrings(sA, sB);
        String tA = cleanNumber(aligned[0].replace(String.valueOf(decimalSeparator), ""));
        String tB = cleanNumber(aligned[1].replace(String.valueOf(decimalSeparator), ""));
        StringBuilder quotient = new StringBuilder();
        StringBuilder current = new StringBuilder();
        int precision = 0;
        boolean hasDecimal = false;
        for (int i = 0; i < tA.length() + 10; i++) {
            if (i < tA.length()) current.append(tA.charAt(i));
            else {
                if (!hasDecimal) { if (quotient.length() == 0) quotient.append("0"); quotient.append(decimalSeparator); hasDecimal = true; }
                current.append('0'); precision++;
            }
            String sCurr = cleanNumber(current.toString());
            int count = 0;
            while (compareAbs(sCurr, tB) >= 0) { sCurr = internalSub(sCurr, tB); count++; }
            current = new StringBuilder(sCurr);
            if (quotient.length() > 0 || count > 0 || i == tA.length() - 1) quotient.append(count);
            if (i >= tA.length() - 1 && cleanNumber(current.toString()).equals("0")) break;
            if (precision >= 10) break;
        }
        String finalQuot = (resNeg && !quotient.toString().equals("0")) ? "-" + quotient : quotient.toString();
        return cleanNumber(finalQuot);
    }

    private static int compareAbs(String sA, String sB) {
        sA = cleanNumber(sA); sB = cleanNumber(sB);
        String[] aligned = alignStrings(sA, sB);
        return aligned[0].compareTo(aligned[1]);
    }
}

File: ReadyToShare/sMath/CSharp/sMath.cs
using System;
using System.Text;
using System.Linq;

namespace JTRSoftware.Math
{
    public static class SMath
    {
        public static char DecimalSeparator = '.';

        private static string GetAbs(string sV) 
            => sV.StartsWith("-") ? sV.Substring(1) : sV;

        private static bool IsNegative(string sV) 
            => sV.StartsWith("-");

        public static string CleanNumber(string sV)
        {
            bool neg = IsNegative(sV);
            string val = GetAbs(sV);

            if (string.IsNullOrEmpty(val)) return "0";

            while (val.Length > 1 && val[0] == '0' && val[1] != DecimalSeparator)
                val = val.Substring(1);

            int dotPos = val.IndexOf(DecimalSeparator);
            if (dotPos >= 0)
            {
                while (val.Length > dotPos + 1 && val[val.Length - 1] == '0')
                    val = val.Substring(0, val.Length - 1);
                if (val[val.Length - 1] == DecimalSeparator)
                    val = val.Substring(0, val.Length - 1);
            }

            if (val == "" || val == DecimalSeparator.ToString()) val = "0";
            return (neg && val != "0") ? "-" + val : val;
        }

        private static void AlignStrings(ref string sA, ref string sB)
        {
            int dotA = sA.IndexOf(DecimalSeparator);
            int dotB = sB.IndexOf(DecimalSeparator);

            int decA = (dotA < 0) ? 0 : sA.Length - dotA - 1;
            int decB = (dotB < 0) ? 0 : sB.Length - dotB - 1;

            while (decA < decB)
            {
                if (dotA < 0) { sA += DecimalSeparator; dotA = sA.Length - 1; }
                sA += "0"; decA++;
            }
            while (decB < decA)
            {
                if (dotB < 0) { sB += DecimalSeparator; dotB = sB.Length - 1; }
                sB += "0"; decB++;
            }

            int intA = (dotA < 0) ? sA.Length : dotA;
            int intB = (dotB < 0) ? sB.Length : dotB;

            while (intA < intB) { sA = "0" + sA; intA++; }
            while (intB < intA) { sB = "0" + sB; intB++; }
        }

        public static string Sum(string sA, string sB)
        {
            if (IsNegative(sA) && IsNegative(sB))
                return "-" + InternalSum(GetAbs(sA), GetAbs(sB));
            if (IsNegative(sA))
                return InternalSub(GetAbs(sB), GetAbs(sA));
            if (IsNegative(sB))
                return InternalSub(GetAbs(sA), GetAbs(sB));

            return CleanNumber(InternalSum(sA, sB));
        }

        private static string InternalSum(string sA, string sB)
        {
            AlignStrings(ref sA, ref sB);
            var res = new StringBuilder();
            int carry = 0;

            for (int i = sA.Length - 1; i >= 0; i--)
            {
                if (sA[i] == DecimalSeparator)
                {
                    res.Insert(0, DecimalSeparator);
                    continue;
                }
                int digitSum = (sA[i] - '0') + (sB[i] - '0') + carry;
                carry = digitSum / 10;
                res.Insert(0, (digitSum % 10).ToString());
            }
            if (carry > 0) res.Insert(0, carry.ToString());
            return res.ToString();
        }

        public static string Sub(string sA, string sB)
        {
            if (IsNegative(sB)) return Sum(sA, GetAbs(sB));
            if (IsNegative(sA)) return "-" + InternalSum(GetAbs(sA), sB);
            return CleanNumber(InternalSub(sA, sB));
        }

        private static string InternalSub(string sA, string sB)
        {
            AlignStrings(ref sA, ref sB);
            bool neg = false;
            if (string.Compare(sA, sB) < 0)
            {
                string tmp = sA; sA = sB; sB = tmp; neg = true;
            }

            var res = new StringBuilder();
            int borrow = 0;

            for (int i = sA.Length - 1; i >= 0; i--)
            {
                if (sA[i] == DecimalSeparator)
                {
                    res.Insert(0, DecimalSeparator);
                    continue;
                }
                int diff = (sA[i] - '0') - (sB[i] - '0') - borrow;
                if (diff < 0) { diff += 10; borrow = 1; }
                else borrow = 0;
                res.Insert(0, diff.ToString());
            }
            string sRes = res.ToString();
            return neg ? "-" + sRes : sRes;
        }

        public static string Multiply(string sA, string sB)
        {
            bool resNeg = IsNegative(sA) ^ IsNegative(sB);
            sA = GetAbs(sA); sB = GetAbs(sB);

            int dotA = sA.IndexOf(DecimalSeparator);
            int decA = (dotA < 0) ? 0 : sA.Length - dotA - 1;
            string intA = sA.Replace(DecimalSeparator.ToString(), "");

            int dotB = sB.IndexOf(DecimalSeparator);
            int decB = (dotB < 0) ? 0 : sB.Length - dotB - 1;
            string intB = sB.Replace(DecimalSeparator.ToString(), "");

            int totalDecs = decA + decB;
            string finalRes = "0";

            for (int i = intB.Length - 1; i >= 0; i--)
            {
                int digitB = intB[i] - '0';
                if (digitB == 0) continue;
                var currentRes = new StringBuilder();
                int carry = 0;
                for (int j = intA.Length - 1; j >= 0; j--)
                {
                    int digitA = intA[j] - '0';
                    int prod = (digitA * digitB) + carry;
                    carry = prod / 10;
                    currentRes.Insert(0, (prod % 10).ToString());
                }
                if (carry > 0) currentRes.Insert(0, carry.ToString());
                currentRes.Append(new string('0', intB.Length - 1 - i));
                finalRes = InternalSum(finalRes, currentRes.ToString());
            }

            if (totalDecs > 0)
            {
                while (finalRes.Length <= totalDecs) finalRes = "0" + finalRes;
                finalRes = finalRes.Insert(finalRes.Length - totalDecs, DecimalSeparator.ToString());
            }

            string result = resNeg ? "-" + finalRes : finalRes;
            return CleanNumber(result);
        }

        public static string Divide(string sA, string sB)
        {
            if (CleanNumber(sB) == "0") return "0";
            bool resNeg = IsNegative(sA) ^ IsNegative(sB);
            sA = GetAbs(sA); sB = GetAbs(sB);

            string tA = sA, tB = sB;
            AlignStrings(ref tA, ref tB);
            tA = tA.Replace(DecimalSeparator.ToString(), "");
            tB = tB.Replace(DecimalSeparator.ToString(), "");
            tA = CleanNumber(tA); tB = CleanNumber(tB);

            string quotient = "";
            string current = "";
            int precision = 0;
            bool hasDecimal = false;

            for (int i = 0; i < tA.Length + 10; i++)
            {
                if (i < tA.Length) current += tA[i];
                else
                {
                    if (!hasDecimal)
                    {
                        if (quotient == "") quotient = "0";
                        quotient += DecimalSeparator;
                        hasDecimal = true;
                    }
                    current += "0";
                    precision++;
                }

                current = CleanNumber(current);
                int count = 0;
                while (CompareAbs(current, tB) >= 0)
                {
                    current = InternalSub(current, tB);
                    count++;
                }

                if (quotient != "" || count > 0 || i == tA.Length - 1)
                    quotient += count.ToString();

                if (i >= tA.Length - 1 && CleanNumber(current) == "0") break;
                if (precision >= 10) break;
            }

            string finalQuot = (resNeg && quotient != "0") ? "-" + quotient : quotient;
            return CleanNumber(finalQuot);
        }

        private static int CompareAbs(string sA, string sB)
        {
            sA = CleanNumber(sA); sB = CleanNumber(sB);
            AlignStrings(ref sA, ref sB);
            return string.Compare(sA, sB);
        }
    }
}
